File: linuxsocket/echo-server.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"USAGE: {progname} <port>"

#from SocketServer import BaseRequestHandler, TCPServer
#import sys, socket
#
#class EchoHandler(BaseRequestHandler):
#    def handle(self):
#        print "Client connected:", self.client_address
#        self.request.sendall(self.request.recv(2**16))
#        self.request.close()
#
#if len(sys.argv) != 2:
#    print __doc__.format(progname=sys.argv[0])
#else:
#    TCPServer(('', int(sys.argv[1])), EchoHandler).serve_forever()

from socket import *
import sys

BLOCKSIZE = 32

def handleConnection(client_sock):
    while True:
        data = client_sock.recv(BLOCKSIZE)
        if not data:
            client_sock.close()
            break
        client_sock.sendall(data)

if len(sys.argv) != 2:
    print __doc__.format(progname=sys.argv[0])
    sys.exit(0)

server_sock = socket(AF_INET, SOCK_STREAM)
server_sock.bind(('', int(sys.argv[1])))
server_sock.listen(5)
while True:
    client_sock, client_addr = server_sock.accept()
    print "Client connected:", client_addr
    handleConnection(client_sock)

File: net-program/linux-socket/echo-server.c
#include <stdio.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <netinet/in.h>

#define MAXPENDING 5        /* Max connection requests */
#define BUFSIZE 32

void Die(char *message) { perror(message); exit(1); }

void handleClient(int sock)
{
    char buf[BUFSIZE];
    int received = -1;
    while ((received = recv(sock, buf, sizeof(buf), 0))) {
        if (received < 0) {
            Die("Failed to receive bytes from client");
        }
        if (send(sock, buf, received, 0) != received) {
            Die("Failed to send bytes to client");
        }
    }
    close(sock);
}

int main(int argc, char *argv[])
{
    int serversock, clientsock;
    struct sockaddr_in echoserver, echoclient;

    if (argc != 3) {
        fprintf(stderr, "USAGE: %s <address> <port>\n", argv[0]);
        return 1;
    }

    /* Create the TCP socket */
    if ((serversock = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
        Die("Failed to create socket");
    }

    /* Construct the server sockaddr_in structure */
    memset(&echoserver, 0, sizeof(echoserver));          /* Clear the structure */
    echoserver.sin_family = AF_INET;                    /* Internet/IP */
    echoserver.sin_addr.s_addr = inet_addr(argv[1]);    /* Incoming addr */
    echoserver.sin_port = htons(atoi(argv[2]));         /* Server port */

    /* Bind the server socket */
    if (bind(serversock,
             (struct sockaddr *) &echoserver,
             sizeof(echoserver)) < 0) {
        Die("Failed to bind the server socket");
    }

    /* Listen on the server socket */
    if (listen(serversock, MAXPENDING) < 0) {
        Die("Failed to listen on server socket");
    }

    /* Run until cancelled */
    while (1) {
        unsigned int clientlen = sizeof(echoclient);
        /* Waiting for client connection */
        if ((clientsock = accept(serversock,
                                 (struct sockaddr *) &echoclient,
                                 &clientlen)) < 0) {
            Die("Failed to accept client connection");
        }
        fprintf(stdout, "Client connected: %s\n",
                inet_ntoa(echoclient.sin_addr));
        handleClient(clientsock);
    }
}
